helm: Add automatic lookup option for k8s service info

kreeuwijk opened the pull request:
This commit adds an option for kubeadm-based clusters to specify
```
k8sServiceHost: "auto"
```
and have a helper function automatically determine the Kubernetes host and port information from the `cluster-info` ConfigMap in the `kube-public` namespace, and use that to populate the `k8sServiceHost` and `k8sServicePort` values. This can significantly simplify the enabling of kubeproxy-less mode, especially for CAPI clusters where the value for `k8sServiceHost` cannot always be known ahead of time.

If any other value than "auto" is provided for `k8sServiceHost`, the behavior reverts to its original form.

```release-note
Add option to automatically discover k8sServiceHost and k8sServicePort info (kubeadm clusters only)
```

gandro commented on install/kubernetes/cilium/templates/_helpers.tpl:
```diff
@@ -165,3 +165,23 @@ Convert a map to a comma-separated string: key1=value1,key2=value2
+  {{- if eq .host "auto" }}
+    {{- $configmap := (lookup "v1" "ConfigMap" "kube-public" "cluster-info") }}
+    {{- $kubeconfig := get $configmap.data "kubeconfig" }}
+    {{- range ( split "\n" $kubeconfig) }}
```
Instead of parsing YAML with "split" and "contains", we could use Helm's `fromYaml` helper here to extract the server:

```
  {{- $k8sServer := get ($kubeconfig | fromYaml) "clusters" | mustFirst | dig "cluster" "server" "" }}
```

kreeuwijk replied:
that's a good idea, I'll test that and adjust the code

gandro commented on install/kubernetes/cilium/templates/_helpers.tpl:
```diff
@@ -165,3 +165,23 @@ Convert a map to a comma-separated string: key1=value1,key2=value2
+      {{- end }}
+    {{- end }}
+  {{- else }}
+    {{- if eq .lookup "host" }}{{ .host | quote }}{{ end }}
```
I think it would be cleaner if we split this out into two separate helpers: `k8sServerHost` and `k8sServerPort`. 

We can then also pass in the full `.` at call sites like this, instead of building and then parsing a custom dict argument:

```
{{ include "k8sServiceHost" . }}
```
or
```
{{ include "k8sServicePort" . }}
```

kreeuwijk replied:
That's fair, I'll look at that

gandro reviewed (commented):
Thank you for your contribution, this is a neat idea! 

Overall, I like the approach but I have a bit of feedback on the implementation

gandro reviewed (commented):
Awesome, thank you, this looks good to me now overall. Two minor things:

1. Please squash your two commits into one
2. I believe you have to run `make -C Documentation update-helm-values` to also update the Helm values in the web docs

sayboras commented on install/kubernetes/cilium/templates/_helpers.tpl:
```diff
@@ -165,3 +165,33 @@ Convert a map to a comma-separated string: key1=value1,key2=value2
+  {{- else }}
+    {{- .Values.k8sServicePort | quote }}
+  {{- end }}
+{{- end }}
```
nit: newline

kreeuwijk replied:
Fixed

sayboras reviewed (commented):
Seems like Documentation/helm-values.tmp is added unintentionally, I have only one nitpick, the rest looks good to me.

sayboras reviewed (approved):
Thanks a lot, lgtm.

kreeuwijk committed:
k8sService: automatic lookup option

This commit adds an option for kubeadm-based clusters to specify
```
k8sServiceHost: "auto"
```
and have a helper function automatically determine the Kubernetes host and port information from the `cluster-info` ConfigMap in the `kube-public` namespace, and use that to populate the `k8sServiceHost` and `k8sServicePort` values. This can significantly simplify the enabling of kubeproxy-less mode, especially for CAPI clusters where the value for `k8sServiceHost` cannot always be known ahead of time.

If any other value than "auto" is provided for `k8sServiceHost`, the behavior reverts to its original form.

Signed-off-by: Kevin Reeuwijk <kevin.reeuwijk@spectrocloud.com>

qmonnet reviewed (approved):
Thank you
